core/src/main/java/com/linecorp/armeria/common/ByteArrayHttpData.java

```diff
@@ -13,201 +13,82 @@
             return this;
         }
-
         if (isEmpty()) {
             return endOfStream ? EMPTY_EOS : EMPTY;
         }
 
-        return new ByteArrayHttpData(array, endOfStream);
-    }
-
-    @Override
-    public boolean isPooled() {
-        return false;
-    }
-
-    @Override
-    public ByteBuf byteBuf(ByteBufAccessMode mode) {
-        requireNonNull(mode, "mode");
-        if (isEmpty()) {
-            return Unpooled.EMPTY_BUFFER;
-        }
-
-        if (mode != ByteBufAccessMode.FOR_IO) {
-            return Unpooled.wrappedBuffer(array);
-        } else {
-            final ByteBuf copy = newDirectByteBuf();
-            copy.writeBytes(array);
-            return copy;
-        }
+        return new ByteArrayHttpData(array(), endOfStream);
     }
 
     @Override
-    public ByteBuf byteBuf(int offset, int length, ByteBufAccessMode mode) {
-        requireNonNull(mode, "mode");
-        if (length == 0) {
-            return Unpooled.EMPTY_BUFFER;
-        }
-
-        if (mode != ByteBufAccessMode.FOR_IO) {
-            return Unpooled.wrappedBuffer(array, offset, length);
-        } else {
-            final ByteBuf copy = newDirectByteBuf(length);
-            copy.writeBytes(array, offset, length);
-            return copy;
-        }
-    }
-
-    private ByteBuf newDirectByteBuf() {
-        return newDirectByteBuf(length());
-    }
-
-    private static ByteBuf newDirectByteBuf(int length) {
-        return PooledByteBufAllocator.DEFAULT.directBuffer(length);
+    public boolean isEndOfStream() {
+        return endOfStream;
     }
 
-    @Override
-    public void close() {}
-
+    @SuppressWarnings("RedundantMethodOverride")
     @Override
     public int hashCode() {
-        // Calculate the hash code from the first 32 bytes.
-        final int end = Math.min(32, length());
-
-        int hash = 0;
-        for (int i = 0; i < end; i++) {
-            hash = hash * 31 + array[i];
-        }
-        return hash;
+        // Use hashcode in ByteBufBytes because we don't use endOfStream in equals.
+        return super.hashCode();
     }
 
     @Override
-    public boolean equals(Object obj) {
-        if (!(obj instanceof HttpData)) {
-            return false;
-        }
-
-        if (obj == this) {
+    public boolean equals(Object o) {
+        if (this == o) {
             return true;
         }
+        if (!(o instanceof HttpData)) {
+            return false;
+        }
 
-        final HttpData that = (HttpData) obj;
+        final HttpData that = (HttpData) o;
         if (length() != that.length()) {
             return false;
         }
 
-        return Arrays.equals(array, that.array());
+        return Arrays.equals(array(), that.array());
+    }
+
+    @Override
+    public String toString() {
+        final String toString = super.toString();
+        if (!isEndOfStream()) {
+            return toString;
+        }
+        return "{EOS}, " + toString;
```

Comment:
nit: place EOS after `toString`? Users might be less interested in EOS than the actual content.